src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -45,6 +46,24 @@
      jruby-puppet jruby-service {:request (dissoc request :ssl-client-cert)}
      (handler (assoc request :jruby-instance jruby-puppet)))))
 
+(defn wrap-with-concurrent-request-limit
+  "Middleware fn that returns a 503 'Service Unavailable' response if the
+  number of outstanding requests for a JRuby instance exceed the given limit.
+  If splay-limit is set, the response will include a Retry-After header set
+  to a random fraction of the limit."
+  [handler metrics-service request-limit splay-limit]
+  ;; TODO: Log warnings when limit is hit. Add i18n to user-facing strings.
+  (fn [request]
+    (let [metrics-map (jruby-metrics/get-metrics metrics-service)
+          requested-count (-> metrics-map :requested-count .getCount)]
+      (if (> requested-count request-limit)
+        (let [response (ringutils/plain-response 503 "JRuby capacity exceeded.")]
+          (if (> splay-limit 0)
+            (assoc-in response [:headers "Retry-After"]
```

Comment:
`average-wait-time` might be difficult threshold to use for when to start rejecting requests with a 503, assuming that once you start rejecting you wouldn't be allowing any more borrow attempts through and so the `average-wait-time` would stay persistently above the threshold.  Using it as a factor in computing what value to return for the `Retry-After` value which accompanies a 503 may make sense, though.

In any event, I agree that starting with an implementation based on the active request count for the threshold to start returning 503 responses and that using the `Retry-After` header randomized from a more static value seems reasonable for now.